### ISSUE TITLE
Memory Leak

### ISSUE THREAD

**Issue by zlshames** [codebytere/node-mac-contacts#33]
## Bug

When calling `contacts.getAllContacts` in repeated succession (every 10 seconds; for testing purposes), memory increases indefinitely.

## Background

Recently, within a side-project of mine, I implemented this library to load contacts from the MacOS AddressBook. Previously, I was accessing the AddressBook database directly. This caused issues because it only accounted for _local_ contacts and not iCloud ones as well. This library helped solve it.

The library is used for API calls as well as matching contact names to phone numbers for a dashboard widget I put together. In this case, the widget would request all contacts every 60 seconds. Now, you may disagree that that's the proper way to use the library, but that's kinda besides the point. So, this widget was requesting a the contacts every 60 seconds, and eventually I saw that the memory of my app drastically increased (> 6 gigs of memory). After hours and hours of debugging, I found out that the call to `getAllContacts` was causing it. When commented out, no memory leak. When called, memory leak.

After noticing that memory was increasing, I set the interval to refresh the dashboard widget to every 10 seconds to see if it would raise the memory faster. Turns out, it did. Memory increased much faster than before (one might say 6x faster).

Now, the question is why is this happening?

## Possible Workaround?

I'm thinking a possible workaround would be to just call `getAllContacts` once, and cache it for later. Then also listen for changes (additions/removals) to add to the cache.

**Comment by codebytere** [comment]
i'll investigate within the week - thanks for raising this!

**Comment by KishanBagaria** [comment]
@zlshames minimal repro code would be useful, will also confirm if it's just library code causing it.

**Comment by zlshames** [comment]
> @zlshames minimal repro code would be useful, will also confirm if it's just library code causing it.

I was planning to add one, just havent had the time. I need to fix and release a version of my app before I can work on trying to show the min reprod code. I just wanted to log it here first because I definitely know it's from the library, just not exactly where within the library.

It should just be a matter of putting the getAllContacts call in a loop and waiting.

**Comment by zlshames** [comment]
Alright, here is my minimum code:

```js
const contacts = require('node-mac-contacts');
const process = require('process');

let count = 0;
setInterval(() => {
    contacts.getAllContacts();

    count += 1;
    const used = process.memoryUsage();
    console.log(
        `Fetching Contacts... Attempt: ${count}; Memory: ${Math.round(used.rss / 1024 / 1024 * 100) / 100} MB`);
}, 3000);
```

And here is the output:
```
Fetching Contacts... Attempt: 1; Memory: 45.11 MB
Fetching Contacts... Attempt: 2; Memory: 49.26 MB
Fetching Contacts... Attempt: 3; Memory: 53.59 MB
Fetching Contacts... Attempt: 4; Memory: 58.59 MB
Fetching Contacts... Attempt: 5; Memory: 62.41 MB
......
......
Fetching Contacts... Attempt: 363; Memory: 614.06 MB
Fetching Contacts... Attempt: 364; Memory: 616.79 MB
Fetching Contacts... Attempt: 365; Memory: 620.51 MB
Fetching Contacts... Attempt: 366; Memory: 622.68 MB
Fetching Contacts... Attempt: 367; Memory: 626.65 MB
Fetching Contacts... Attempt: 368; Memory: 612.72 MB
Fetching Contacts... Attempt: 369; Memory: 615.69 MB
Fetching Contacts... Attempt: 370; Memory: 620.18 MB
Fetching Contacts... Attempt: 371; Memory: 621.08 MB
Fetching Contacts... Attempt: 372; Memory: 624.26 MB
Fetching Contacts... Attempt: 373; Memory: 627.69 MB
Fetching Contacts... Attempt: 374; Memory: 628.63 MB
Fetching Contacts... Attempt: 375; Memory: 610.58 MB
Fetching Contacts... Attempt: 376; Memory: 611.84 MB
Fetching Contacts... Attempt: 377; Memory: 615.65 MB
Fetching Contacts... Attempt: 378; Memory: 621.02 MB
Fetching Contacts... Attempt: 379; Memory: 588.06 MB
Fetching Contacts... Attempt: 380; Memory: 592.41 MB
Fetching Contacts... Attempt: 381; Memory: 594.72 MB
Fetching Contacts... Attempt: 382; Memory: 595.98 MB
Fetching Contacts... Attempt: 383; Memory: 599.43 MB
Fetching Contacts... Attempt: 384; Memory: 602.69 MB
Fetching Contacts... Attempt: 385; Memory: 606.75 MB
Fetching Contacts... Attempt: 386; Memory: 590.24 MB
Fetching Contacts... Attempt: 387; Memory: 591.64 MB
Fetching Contacts... Attempt: 388; Memory: 594.68 MB
Fetching Contacts... Attempt: 389; Memory: 599.36 MB
Fetching Contacts... Attempt: 390; Memory: 602.18 MB
Fetching Contacts... Attempt: 391; Memory: 604.71 MB
Fetching Contacts... Attempt: 392; Memory: 608.7 MB
Fetching Contacts... Attempt: 393; Memory: 612 MB
Fetching Contacts... Attempt: 394; Memory: 588.04 MB
Fetching Contacts... Attempt: 395; Memory: 587.16 MB
```

I cut out the middle part to reduce logs. Needless to say, it definitely is leaking memory somehow... You can see memory actually reducing at some points, but I think that's just from the garbage collector, and unrelated to the app.

I do also get a handful of warnings each time i pull from my contacts. Here is an example:

```
CoreData: error: Failed to create safeguard location for file: /Users/xxxxxxxx/Library/Application Support/AddressBook/.AddressBook-v22_SUPPORT/_EXTERNAL_DATA/7E653E7C-047C-4FC5-8DBB-FFC22A6A5CED error: Error Domain=NSCocoaErrorDomain Code=4 "The file “7E653E7C-047C-4FC5-8DBB-FFC22A6A5CED” doesn’t exist." UserInfo={NSSourceFilePathErrorKey=/Users/xxxxxxxx/Library/Application Support/AddressBook/.AddressBook-v22_SUPPORT/_EXTERNAL_DATA/7E653E7C-047C-4FC5-8DBB-FFC22A6A5CED, NSUserStringVariant=(
    Copy
), NSDestinationFilePath=/var/folders/lb/syt5h7fd2gn89h4z1g84qscw0000gp/T/.LINKS/AD00AEC7-C204-4AFD-B43C-5CA78CB8933A/7E653E7C-047C-4FC5-8DBB-FFC22A6A5CED_0x7fb367ed6830, NSFilePath=/Users/xxxxxxxx/Library/Application Support/AddressBook/.AddressBook-v22_SUPPORT/_EXTERNAL_DATA/7E653E7C-047C-4FC5-8DBB-FFC22A6A5CED, NSUnderlyingError=0x7fb367ed7f80 {Error Domain=NSPOSIXErrorDomain Code=2 "No such file or directory"}}
```

I'm not entirely sure if these are ignorable, but I have always thought they were

**Comment by KishanBagaria** [comment]
Think you might wanna try forcing GC with `require('v8').setFlagsFromString('--expose_gc'); require('vm').runInNewContext('gc')()`.

> CoreData: error: Failed to create safeguard location for file

That's an Apple thing afaik, also happens with native apps.

**Comment by zlshames** [comment]
I tried the same reproduction code as above, just with a `gc` call after each fetch of contacts.

```js
const contacts = require('node-mac-contacts');
const process = require('process');

require('v8').setFlagsFromString('--expose_gc');

let count = 0;
setInterval(() => {
    contacts.getAllContacts();
    require('vm').runInNewContext('gc')();

    count += 1;
    const used = process.memoryUsage();
    console.log(
        `Fetching Contacts... Attempt: ${count}; Memory: ${Math.round(used.rss / 1024 / 1024 * 100) / 100} MB`);
}, 3000);
```

However, same/similar result:

```
Fetching Contacts... Attempt: 1; Memory: 46.94 MB
Fetching Contacts... Attempt: 2; Memory: 52.05 MB
Fetching Contacts... Attempt: 3; Memory: 55.79 MB
Fetching Contacts... Attempt: 4; Memory: 59.37 MB
Fetching Contacts... Attempt: 5; Memory: 63.65 MB
Fetching Contacts... Attempt: 6; Memory: 66.25 MB
Fetching Contacts... Attempt: 7; Memory: 69.16 MB
Fetching Contacts... Attempt: 8; Memory: 72.23 MB
Fetching Contacts... Attempt: 9; Memory: 75.83 MB
Fetching Contacts... Attempt: 10; Memory: 78.9 MB
Fetching Contacts... Attempt: 11; Memory: 83.01 MB
Fetching Contacts... Attempt: 12; Memory: 86.95 MB
Fetching Contacts... Attempt: 13; Memory: 91.01 MB
Fetching Contacts... Attempt: 14; Memory: 93.01 MB
Fetching Contacts... Attempt: 15; Memory: 95.67 MB
Fetching Contacts... Attempt: 16; Memory: 99.43 MB
Fetching Contacts... Attempt: 17; Memory: 103.41 MB
Fetching Contacts... Attempt: 18; Memory: 107.48 MB
Fetching Contacts... Attempt: 19; Memory: 110.59 MB
Fetching Contacts... Attempt: 20; Memory: 113.04 MB
Fetching Contacts... Attempt: 21; Memory: 117.87 MB
Fetching Contacts... Attempt: 22; Memory: 121.14 MB
Fetching Contacts... Attempt: 23; Memory: 124.53 MB
Fetching Contacts... Attempt: 24; Memory: 126.25 MB
```

Memory seems to just keep rising... Please let me know if I implemented the calls incorrectly.

**Comment by codebytere** [comment]
Thanks! i'll start investigating :)

**Comment by codebytere** [comment]
I'm not sure it's a memory leak per se, but i did narrow it down to this call: https://github.com/codebytere/node-mac-contacts/blob/ea3a34d6231bdf3882b01f3a97f9dd20d4048b5c/contacts.mm#L516-L519

**Comment by codebytere** [comment]
Ok - unfortunately- this isn't a memory leak. This is, unfortunately, just an expensive API on Apple's end. I'm playing with using an enumerator, as [docs say that's less expensive](https://developer.apple.com/documentation/contacts/cncontactstore/3294191-enumeratorforcontactfetchrequest), but even using an enumerator and wrapping the call in an autorelease pool:

```diff
diff --git a/contacts.mm b/contacts.mm
index 42843f3..a03c038 100644
--- a/contacts.mm
+++ b/contacts.mm
@@ -509,14 +509,17 @@ Napi::Array GetAllContacts(const Napi::CallbackInfo &info) {
   // This is a set so that contacts in multiple containers aren't duplicated.
   NSMutableSet *unordered_contacts = [[NSMutableSet alloc] init];
   int num_containers = [containers count];
+  CNContactFetchRequest *request = [[CNContactFetchRequest alloc] initWithKeysToFetch:GetContactKeys(extra_keys)];
   for (int idx = 0; idx < num_containers; idx++) {
-    CNContainer *container = [containers objectAtIndex:idx];
-    NSPredicate *predicate = [CNContact
-        predicateForContactsInContainerWithIdentifier:[container identifier]];
-    NSArray *container_contacts =
-        [addressBook unifiedContactsMatchingPredicate:predicate
-                                          keysToFetch:GetContactKeys(extra_keys)
+    NSAutoreleasePool *pool = [[NSAutoreleasePool alloc] init];
+    CNFetchResult *fetch_result = [addressBook enumeratorForContactFetchRequest:request
                                           error:&error];
+    NSEnumerator *enumerator = [fetch_result value];
+    id object;
+    while ((object = [enumerator nextObject])) {
+      [unordered_contacts addObject:object];
+    }
+    [pool drain];
     if (error != nil) {
       std::string err_msg =
           std::string([error.localizedDescription UTF8String]);
@@ -524,8 +527,6 @@ Napi::Array GetAllContacts(const Napi::CallbackInfo &info) {
           .ThrowAsJavaScriptException();
       return Napi::Array::New(env);
     }
-
-    [unordered_contacts addObjectsFromArray:container_contacts];
   }
 
   Napi::Array contacts = Napi::Array::New(env);
```

doesn't seem to improve things much:

```
node-mac-contacts on git:main ❯ node test.js                                    10:54AM
2022-06-17 10:55:19.207 node[37206:8699203] XXX: countOfStores: 1, countOfAccounts: 1
Fetching Contacts... Attempt: 1; Memory: 75.06 MB
Fetching Contacts... Attempt: 2; Memory: 85.55 MB
Fetching Contacts... Attempt: 3; Memory: 93.89 MB
```

happy to keep poking at it though

**Comment by zlshames** [comment]
Thank you for looking into this! Very interesting finds... Though, a bit unfortunate. I understand that they can be expensive calls, but how come the memory isn't being freed? For instance, in our test case we check VERY quickly. However, in my real world scenario, I was only checking once a minute or so. Even then, the memory just kept compiling and compiling until it was gigs big. It took much longer, but had the same outcome. 

In either case, I think my solution now has been working well:

- Fetch all contacts on app startup
- Listen for changes to contacts and whenever it changes, we can refresh the list. 

Now, this could theoretically still cause the leak if it detects a TON of refreshes. Seeing how each call will only I crease memory by 3-5mb, it's not a huge deal... Just something to worry a little bit about. An app refresh will always clear the memory, but obviously that is less ideal. I could also remove the listener and just let the user manually refresh whenever. I already do that, but the listener typically removes the manual requirement. 

Either way, thank you for investigating this! If the enumerated stuff isn't too bad to use and maintains functionality, I'm all for it. Otherwise, I guess there isn't a crazy amount of things we can do to get around this. It is what it is. Thanks Apple 🥲

**Comment by KishanBagaria** [comment]
> Listen for changes to contacts and whenever it changes, we can refresh the list.

Debouncing for 1000-3000ms would be a good idea, sometimes change listener fires multiple times for a single change.

**Comment by zlshames** [comment]
> > Listen for changes to contacts and whenever it changes, we can refresh the list.
> 
> Debouncing for 1000-3000ms would be a good idea, sometimes change listener fires multiple times for a single change.

Good idea, I think I've seen instances where this happened, or at least the logs I've seen indicated that it happened. I'll look into it. It's not a huge problem now, so nothing urgent

**Comment by codebytere** [comment]
ha wow good timing bc i am actually in the middle of that! working on caching it as we speak.

**Comment by codebytere** [comment]
Alright - i think https://github.com/codebytere/node-mac-contacts/commit/0412ea6d7ab832263370fd5d7f17ed9db9fb0874 should do it.

tested with the following:
```js
const { listener, getAllContacts, addNewContact } = require('./index')
const process = require('process')

let count = 0

listener.setup();

setInterval(() => {
  getAllContacts()

  count += 1
  const used = process.memoryUsage()
  console.log(
    `Fetching Contacts... Attempt: ${count}; Memory: ${
      Math.round((used.rss / 1024 / 1024) * 100) / 100
    } MB`,
  )
}, 2000)

setTimeout(() => {
  console.log('adding new contact')
  addNewContact({
    firstName: 'William',
    lastName: 'Grapeseed',
    nickname: 'Billy',
    birthday: '1990-09-09',
    phoneNumbers: ['+1234567890'],
    emailAddresses: ['billy@grapeseed.com'],
  })
}, 10000)
```

Output:
```
node-mac-contacts on git:main ❯ node test.js                                     3:25PM
2022-06-17 15:25:04.116 node[78646:8880923] XXX: countOfStores: 1, countOfAccounts: 1
Fetching Contacts... Attempt: 1; Memory: 69.55 MB
Fetching Contacts... Attempt: 2; Memory: 69.94 MB
Fetching Contacts... Attempt: 3; Memory: 70.07 MB
Fetching Contacts... Attempt: 4; Memory: 70.07 MB
adding new contact
Fetching Contacts... Attempt: 5; Memory: 84.7 MB
Fetching Contacts... Attempt: 6; Memory: 84.71 MB
Fetching Contacts... Attempt: 7; Memory: 84.71 MB
Fetching Contacts... Attempt: 8; Memory: 85.27 MB
Fetching Contacts... Attempt: 9; Memory: 85.24 MB
Fetching Contacts... Attempt: 10; Memory: 85.25 MB
Fetching Contacts... Attempt: 11; Memory: 85.25 MB
Fetching Contacts... Attempt: 12; Memory: 85.26 MB
```

i'll see what i can do about clearing existing memory - it might be a little sordid depending on the SDK mechanics but for now it's an improvement.

**Comment by zlshames** [comment]
Wonderful! Thank you!

**Comment by zlshames** [comment]
Let me know if you need anything else from me in order to get this as an official release

**Comment by codebytere** [comment]
Whoops, thanks for the reminder! just published `node-mac-contacts@1.6.0` :) Also going to close this out.
